### PR TITLE
Migrate ModelMultipleChoiceFilters to NaturalKeyOrPKMultipleChoiceFilter 

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes: #1916
# What's Changed

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Comment by bryanculver** [comment]
Double checking here: I thought we rolled back/held off `NaturalKeyOrPKMultpleChoiceFilter` due to: https://github.com/nautobot/nautobot/issues/2875

At glance it seems most of these are OK because they are using the slug field which is acting as a natural key at the moment, is that correct?

**Review by bryanculver — COMMENTED** [review]
Code changes look good, just want to make sure this is not going to essentially re-introduce https://github.com/nautobot/nautobot/pull/2882

**Comment by timizuoebideri1** [comment]
> At glance it seems most of these are OK because they are using the slug field which is acting as a natural key at the moment, is that correct?

Yup, 💯%  correct

**Review by bryanculver — COMMENTED** [review]
LGTM as soon as the tests are passing!

File: nautobot/circuits/filters.py [hunk 1]
```diff
@@ -46,13 +47,12 @@ class ProviderFilterSet(NautobotFilterSet):
     site_id = django_filters.ModelMultipleChoiceFilter(
         field_name="circuits__terminations__site",
         queryset=Site.objects.all(),
-        label="Site",
+        label="Site (ID) - Deprecated (use site filter)",
     )
-    site = django_filters.ModelMultipleChoiceFilter(
-        field_name="circuits__terminations__site__slug",
+    site = NaturalKeyOrPKMultipleChoiceFilter(
+        field_name="circuits__terminations__site",
```
Comment (glennmatthews) [inline nautobot/circuits/filters.py:53]: Is this filter tested/confirmed to work with this change? I can't remember if we have other cases where we're using `NaturalKeyOrPKMultipleChoiceFilter` with a nested lookup like this.

File: nautobot/circuits/filters.py [hunk 2]
```diff
@@ -141,24 +137,21 @@ class CircuitFilterSet(NautobotFilterSet, StatusModelFilterSetMixin, TenancyMode
...
-        field_name="terminations__site__slug",
+    site = NaturalKeyOrPKMultipleChoiceFilter(
+        field_name="terminations__site",
         queryset=Site.objects.all(),
-        to_field_name="slug",
-        label="Site (slug)",
+        label="Site (ID or slug)",
     )
     location = TreeNodeMultipleChoiceFilter(
         field_name="terminations__location__slug",
```
Comment (glennmatthews) [inline nautobot/circuits/filters.py:157]: I think we could also mark `region_id` (below) as deprecated?

File: nautobot/docs/development/best-practices.md [hunk 3]
```diff
@@ -197,15 +197,9 @@ class UserFilter(NautobotFilterSet):
     - A PK-based filter **should** be made available as well, generally with a name suffixed by `_id`. For example:
```
Comment (glennmatthews) [inline nautobot/docs/development/best-practices.md:197]: I think this instruction is no longer accurate, given the change to the below example? Please review.

File: nautobot/circuits/filters.py (hunk 1 above)
Comment (timizuoebideri1) [reply nautobot/circuits/filters.py:53]: Yup, Its been tested.

File: nautobot/circuits/tests/test_filters.py [hunk 4]
```diff
@@ -94,7 +99,7 @@ class CircuitTestCase(FilterTestCases.FilterTestCase, FilterTestCases.TenancyFil
     @classmethod
     def setUpTestData(cls):
 
-        cls.regions = Region.objects.filter(sites__isnull=False).distinct()[:3]
+        cls.regions = Region.objects.filter(sites__isnull=False).distinct()
```
Comment (glennmatthews) [inline nautobot/circuits/tests/test_filters.py:102]: Why this change and similar ones below?

File: nautobot/extras/tests/test_filters.py [hunk 5]
```diff
@@ -189,6 +196,9 @@ def setUpTestData(cls):
         )
         cls.platforms = platforms
 
+        locations = Location.objects.all()[:3]
+        cls.locations = locations
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_filters.py:200]:
```suggestion
        cls.locations = Location.objects.all()[:3]
```

File: nautobot/users/filters.py [hunk 6]
```diff
@@ -32,8 +32,9 @@ class UserFilterSet(BaseFilterSet):
     group_id = django_filters.ModelMultipleChoiceFilter(
         field_name="groups",
         queryset=Group.objects.all(),
-        label="Group",
+        label="Group (ID)",
     )
+    # TODO(timizuo): Migrate ModelMultipleChoiceFilter to NaturalKeyOrPKMultipleChoiceFilter
```
Comment (glennmatthews) [inline nautobot/users/filters.py:37]: IIRC this is blocked because NaturalKeyOrPKMultipleChoiceFilter isn't correctly handling integer inputs?

**Review by glennmatthews — APPROVED** [review]
Looks pretty solid to me. Thanks!

File: nautobot/circuits/tests/test_filters.py (hunk 4 above)
Comment (timizuoebideri1) [reply nautobot/circuits/tests/test_filters.py:102]: Was trying to fix there MySQL error

File: nautobot/users/filters.py (hunk 6 above)
Comment (timizuoebideri1) [reply nautobot/users/filters.py:37]: Yup

File: nautobot/circuits/tests/test_filters.py (hunk 4 above)
Comment (glennmatthews) [reply nautobot/circuits/tests/test_filters.py:102]: I think the approach we've used elsewhere to work properly with MySQL is explicitly casting the sliced queryset to a list, e.g.:

```suggestion
        cls.regions = list(Region.objects.filter(sites__isnull=False).distinct()[:3])
```

File: nautobot/ipam/tests/test_filters.py [hunk 7]
```diff
@@ -1051,17 +1043,19 @@ def test_port(self):
 
     def test_device(self):
         devices = Device.objects.all()[:2]
```
Comment (timizuoebideri1) [inline nautobot/ipam/tests/test_filters.py:1045]:
```suggestion
        devices = list(Device.objects.all()[:2])
```

File: nautobot/ipam/tests/test_filters.py [hunk 8]
```diff
@@ -1051,17 +1043,19 @@ def test_port(self):
...
-        params = {"device": [devices[0].name, devices[1].name]}
-        self.assertEqual(self.filterset(params, self.queryset).qs.count(), 2)
+        filter_params = [{"device_id": [devices[0].pk, devices[1].pk]}, {"device": [devices[0].pk, devices[1].name]}]
+        for params in filter_params:
+            self.assertQuerysetEqualAndNotEmpty(
+                self.filterset(params, self.queryset).qs, self.queryset.filter(device__in=devices).distinct()
+            )
 
     def test_virtual_machine(self):
         vms = VirtualMachine.objects.all()[:2]
```
Comment (timizuoebideri1) [inline nautobot/ipam/tests/test_filters.py:1053]:
```suggestion
        vms = list(VirtualMachine.objects.all()[:2])
```

File: nautobot/users/filters.py (hunk 6 above)
Comment (glennmatthews) [reply nautobot/users/filters.py:37]: Would you mind clarifying this information in the comment?

Comment (timizuoebideri1) [inline nautobot/users/filters.py:37]:
```suggestion
    # TODO(timizuo): Migrate ModelMultipleChoiceFilter to NaturalKeyOrPKMultipleChoiceFilter: As of now NaturalKeyOrPKMultipleChoiceFilter isn't correctly handling integer id field
```
